router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
> You might end up trying to cancel the request before it's begun, and then the request to the backend will still happen.

I think checking `ctx.Done()` (the `select`/`default`) prevents a quickly canceled request from being sent to a backend. But we aren't doing anything (e.g. reading the request body) to trigger that `CloseNotifier`. We've already got the request body wrapped in `fakeCloseReadCloser`, perhaps we should `Peek` right before starting the `for` loop? i think that would trigger the `CloseNotifier` and trickle down to `ctx.Done()` (but across different goroutines, so it's not guaranteed).